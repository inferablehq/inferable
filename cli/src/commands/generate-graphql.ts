import { CommandModule } from "yargs";
import fs from "fs";
import os from "os";
import { mkdir } from "fs/promises";
import * as Handlebars from "handlebars";

import { jsonSchemaToZod } from "json-schema-to-zod";

import { input } from "@inquirer/prompts";

import * as path from "path";
import { fromGraphQL } from "./../lib/graphql";

interface GenerateGraphqlArgs {
  schema?: string;
  operations?: string;
  name?: string;
  dir?: string;
}

export const GenerateGraphql: CommandModule<{}, GenerateGraphqlArgs> = {
  command: "graphql [name] [schema] [operations] [dir]",
  describe: "Generates an Inferable Service from a GraphQL schema",
  builder: (yargs) =>
    yargs
      .positional("name", {
        describe: "Name of the service",
        type: "string",
      })
      .positional("schema", {
        describe: "GraphQL schema file",
        type: "string",
      })
      .positional("operations", {
        describe: "GraphQL operations file",
        type: "string",
      })
      .positional("dir", {
        describe: "Directory to create service in",
        type: "string",
      }),
  handler: async ({ schema, operations, name, dir }) => {
    const tmpDir = `${os.tmpdir()}/${Math.random().toString(36).substring(2, 15)}`;

    await mkdir(tmpDir);
    const distPath = `${tmpDir}`;

    if (!name) {
      name = await input({
        message: "Inferable service name",
        validate: (value) => {
          if (!value) {
            return "Service name is required";
          }
          return true;
        },
      });
    }

    if (name.match(/[^A-Za-z0-9]/)) {
      console.error("Service name can only contain letters and numbers");
      process.exit(1);
    }

    if (!schema) {
      schema = await input({
        message: "GraphQL schema file",
        validate: (value) => {
          if (!value) {
            return "GraphQL schema is required";
          }
          return true;
        },
      });
    }

    if (!operations) {
      operations = await input({
        message: "GraphQL operations file",
        validate: (value) => {
          if (!value) {
            return "GraphQL operations are required";
          }
          return true;
        },
      });
    }

    if (!dir) {
      // Check if ./src/services dir exists
      const proxyServiceDir = "./src/services";
      if (fs.existsSync(proxyServiceDir)) {
        console.log("Detected Proxy project structure");
        dir = proxyServiceDir;
      } else {
        console.log("No directory specified, using current directory");
        dir = "./";
      }
    }

    const serviceFileName = `${name}.service.ts`;
    const tmpServiceFile = path.join(distPath, serviceFileName);

    fs.mkdirSync(path.join(dir, name), { recursive: true });
    fs.mkdirSync(path.join(dir, name, "functions"), { recursive: true });

    generateServiceFile(name, tmpServiceFile);
    fs.renameSync(tmpServiceFile, path.join(dir, name, serviceFileName));

    if (fs.existsSync(path.join(dir, name, `${name}.config.ts`))) {
      console.log(
        `Config file already exists: ${name}.config.ts. Skipping config generation...`,
      );
    } else {
      generateConfigFile(name, path.join(dir, name, `${name}.config.ts`));
    }

    console.log(`✅ Service file generated: ${serviceFileName}`);

    const functions = await fromGraphQL({
      schema,
      operations,
    });

    functions.forEach((f) => {
      const functionfileName = `${name}.${f.name}.ts`;
      const tmpFnFile = path.join(distPath, functionfileName);

      generateFunctionFile(f, name, tmpFnFile);

      // TODO: Check for existing files and warn
      fs.renameSync(
        tmpFnFile,
        path.join(dir, name, "functions", functionfileName),
      );

      console.log(`✅ Function file generated: ${functionfileName}`);
    });
  },
};

const generateConfigFile = (name: string, outPath: string) => {
  const configTemplate = `if (!process.env.INFERABLE_SERVICE_${name.toUpperCase()}_URL) {
  throw new Error("Missing environment variable: INFERABLE_SERVICE_${name.toUpperCase()}_URL");
}

const config = {
  axiosDefaults() {
    return {
      baseURL: process.env.INFERABLE_SERVICE_${name.toUpperCase()}_URL,
      headers: {
        "Content-Type": "application/json",
      },
    };
  },
};

export default config;
`;

  fs.writeFileSync(outPath, configTemplate, "utf8");
};

const generateServiceFile = (name: string, outPath: string) => {
  const serviceTemplate = `import fs from "fs";
import path from "path";

const functions = fs
  .readdirSync(path.join(__dirname, "functions"))
  .filter((f) => {
    const [service, name, ext, ...rest] = f.split(".");
    const isService = service === "{{serviceName}}" && (ext === "ts" || ext === "js") && name && !rest.length;

    if (!isService) {
      console.debug("Skipping function file", f);
    }

    return isService;
  })
  .map((f) => require(\`./functions/\${f}\`).default)
  .filter(Boolean);

if (functions.length === 0) {
  throw new Error("No valid functions found for service {{serviceName}}");
}

export default {
  name: "{{serviceName}}",
  functions,
};
`;

  const compiledTemplate = Handlebars.compile(serviceTemplate);

  const generatedServiceFile = compiledTemplate({
    serviceName: name,
  });

  fs.writeFileSync(outPath, generatedServiceFile, "utf8");
};

const generateFunctionFile = (
  fn: any,
  serviceName: string,
  outPath: string,
) => {
  const isMutation = fn.query.trim().startsWith("mutation");

  const functionTemplate = `import axios from 'axios';
import z from 'zod';
import config from '../{{serviceName}}.config';

const inputSchema = {{zodSchema}}

const query = \`
{{query}}
\`

export default {
  name: "{{name}}",
  description: "{{description}}",
  func: async (input: Zod.infer<typeof inputSchema>) => {

   const result = await axios({
      ...config.axiosDefaults(),
     method: "POST",
     data: {
       query,
       variables: input
     },
    });

    return result.data
  },
  schema: {
    input: inputSchema
  },
  config: {
    requiresApproval: {{requiresApproval}}
  }
}
 `;
  const zodSchema = jsonSchemaToZod(fn.schema.input as any, { noImport: true });

  const compiledTemplate = Handlebars.compile(functionTemplate, {
    noEscape: true,
  });

  const generatedFunctionFile = compiledTemplate({
    ...fn,
    description: fn.description || `Function for the ${fn.name} operation`,
    serviceName,
    zodSchema,
    requiresApproval: isMutation ? "true" : "false",
  });

  fs.writeFileSync(outPath, generatedFunctionFile, "utf8");
};
