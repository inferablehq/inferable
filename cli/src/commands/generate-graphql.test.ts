import { GenerateGraphql } from "./generate-graphql";
import * as path from "path";
import * as os from "os";
import * as fs from "fs";

describe("generate-graphql-api", () => {
  it("should generate graphql api", async () => {
    const random = Math.random().toString(36).substring(2, 15);

    fs.mkdirSync(path.resolve(os.tmpdir(), random));

    await GenerateGraphql.handler({
      schema: path.resolve(process.cwd(), "src", "testdata", "link.schema.gql"),
      operations: path.resolve(
        process.cwd(),
        "src",
        "testdata",
        "link.operations.gql",
      ),
      name: "link",
      dir: path.resolve(os.tmpdir(), random),
    } as any);

    expect(
      fs.existsSync(
        path.resolve(os.tmpdir(), random, "link", "link.config.ts"),
      ),
    ).toBe(true);
    expect(
      fs.existsSync(
        path.resolve(os.tmpdir(), random, "link", "link.service.ts"),
      ),
    ).toBe(true);
    expect(
      fs
        .readFileSync(
          path.resolve(os.tmpdir(), random, "link", "link.config.ts"),
        )
        .toString(),
    ).toMatchSnapshot();
    expect(
      fs
        .readFileSync(
          path.resolve(os.tmpdir(), random, "link", "link.service.ts"),
        )
        .toString(),
    ).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(
          os.tmpdir(),
          random,
          "link",
          "functions",
          "link.MergePullRequest.ts",
        ),
        "utf-8",
      ),
    ).toMatchSnapshot();
  });
});
