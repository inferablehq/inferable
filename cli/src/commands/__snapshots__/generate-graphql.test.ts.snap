// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-graphql-api should generate graphql api 1`] = `
"if (!process.env.INFERABLE_SERVICE_LINK_URL) {
  throw new Error("Missing environment variable: INFERABLE_SERVICE_LINK_URL");
}

const config = {
  axiosDefaults() {
    return {
      baseURL: process.env.INFERABLE_SERVICE_LINK_URL,
      headers: {
        "Content-Type": "application/json",
      },
    };
  },
};

export default config;
"
`;

exports[`generate-graphql-api should generate graphql api 2`] = `
"import fs from "fs";
import path from "path";

const functions = fs
  .readdirSync(path.join(__dirname, "functions"))
  .filter((f) => {
    const [service, name, ext, ...rest] = f.split(".");
    const isService = service === "link" && (ext === "ts" || ext === "js") && name && !rest.length;

    if (!isService) {
      console.debug("Skipping function file", f);
    }

    return isService;
  })
  .map((f) => require(\`./functions/\${f}\`).default)
  .filter(Boolean);

if (functions.length === 0) {
  throw new Error("No valid functions found for service link");
}

export default {
  name: "link",
  functions,
};
"
`;

exports[`generate-graphql-api should generate graphql api 3`] = `
"import axios from 'axios';
import z from 'zod';
import config from '../link.config';

const inputSchema = z.object({ "username": z.string(), "slug": z.string(), "pid": z.string() })

const query = \`
mutation MergePullRequest($username: String!, $slug: String!, $pid: String!) {
  mergePullRequest(username: $username, slug: $slug, pid: $pid)
}
\`

export default {
  name: "MergePullRequest",
  description: "Function for the MergePullRequest operation",
  func: async (input: Zod.infer<typeof inputSchema>) => {

   const result = await axios({
      ...config.axiosDefaults(),
     method: "POST",
     data: {
       query,
       variables: input
     },
    });

    return result.data
  },
  schema: {
    input: inputSchema
  },
  config: {
    requiresApproval: true
  }
}
 "
`;
